lib/bowling.rb

```diff
@@ -0,0 +1,128 @@
+class Bowling
+  attr_reader :knocked_down, :first_roll, :scores , :final_scores, :frame, :spares, :strikes, :extra_turn
+  def initialize
+    @scores = create_score_sheet
+    @final_scores = create_score_sheet
+    @first_roll = true
+    @frame = 1
+    @knocked_down = 0
+    @spares = []
+    @strikes = []
+    @extra_turn = false
+    @tenth_frame_rolls = 0
+  end
+
+  def roll(pins)
+    @knocked_down += pins
+    return tenth_frame(pins) if frame == 10
+    scores[frame] << pins
+    update_bonus_points(pins)
+    
+    if first_roll 
+      return strike if knocked_down == 10
+    else
+      spares << frame if knocked_down == 10
```

Comment:
Rather than keeping a record of strikes and spares, then determining additional points based on what strikes and spares there are, perhaps you could just keep track, for each frame, of how many additional rolls should be factored in?